plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java

```diff
@@ -0,0 +1,1299 @@
+        public byte[] serialize(String topic, SqlVarbinary data)
+        {
+            if (data == null) {
+                return null;
+            }
+            return data.getBytes();
+        }
+    }
+
+    private <T> void testKey(Type keyType, Class<? extends Serializer> keySerializerClass, T firstValue, T secondValue)
+    {
+        String topic = "topic-" + getDisplayNameForTopic(keyType) + "-" + UUID.randomUUID();
+        String tableName = toDoubleQuoted(topic + "&key-column=key_col:" + keyType.getDisplayName());
+        Schema schema = SchemaBuilder.record("test").fields()
+                .name("string_optional_field").type().optional().stringType()
+                .endRecord();
+        testingKafka.sendMessages(Stream.of(new ProducerRecord<T, GenericRecord>(topic, firstValue, new GenericRecordBuilder(schema).build())), schemaRegistryAwareProducer(testingKafka)
+                .put(KEY_SERIALIZER_CLASS_CONFIG, keySerializerClass.getName())
+                .put(VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class.getName())
+                .buildOrThrow());
+        waitUntilTableExists(topic);
+        QueryAssertions queryAssertions = new QueryAssertions(getQueryRunner());
+
+        queryAssertions.query("SELECT key_col, string_optional_field FROM " + tableName)
+                .assertThat()
+                .matches("VALUES (CAST(" + toSingleQuotedOrNullLiteral(firstValue) + " AS " + keyType.getDisplayName() + "), CAST(NULL AS VARCHAR))");
+        assertUpdate("INSERT INTO " + tableName + " (key_col) VALUES" +
+                "  (CAST(" + toSingleQuotedOrNullLiteral(secondValue) + " AS " + keyType.getDisplayName() + "))," +
+                "  (CAST(NULL AS " + keyType.getDisplayName() + "))", 2);
+        queryAssertions.query("SELECT key_col, string_optional_field FROM " + tableName)
+                .assertThat()
+                .matches("VALUES (CAST(" + toSingleQuotedOrNullLiteral(firstValue) + " AS " + keyType.getDisplayName() + "), CAST(NULL AS VARCHAR))," +
+                        "  (CAST(" + toSingleQuotedOrNullLiteral(secondValue) + " AS " + keyType.getDisplayName() + "), CAST(NULL AS VARCHAR))," +
+                        "  (CAST(NULL AS " + keyType.getDisplayName() + "), CAST(NULL AS VARCHAR))");
+    }
+
+    private static String getDisplayNameForTopic(Type type)
+    {
+        return type.getDisplayName().replaceAll("[()]", "_");
+    }
+
+    private String toSingleQuotedOrNullLiteral(Object value)
+    {
+        if (value == null) {
+            return "NULL";
+        }
+        if (value instanceof SqlVarbinary) {
+            return "X'" + BaseEncoding.base16().encode(((SqlVarbinary) value).getBytes()) + "'";
+        }
+        else if (value instanceof byte[]) {
+            return "X'" + BaseEncoding.base16().encode(deserializeBytes((byte[]) value)) + "'";
+        }
+        return "'" + value + "'";
+    }
+
+    @Test
+    public void testSupportedKeyColumnTypes()
```

Comment:
Can we move such low level tests out of the SmokeTest? To separate test class?

Reply:
Extracted to `TestKeyColumnsConnectorSmokeTest`.